lua/plenary/path.lua

```diff
@@ -198,6 +199,54 @@ function Path:expand()
   return expanded and expanded or error("Path not valid")
 end
 
+function Path:make_relative(filepath, cwd)
+  if not cwd or not filepath then return filepath end
+
+  if filepath:sub(1, #cwd) == cwd  then
+    local offset =  0
+    -- if  cwd does ends in the os separator, we need to take it off
+    if cwd:sub(#cwd, #cwd) ~= path.separator then
+      offset = 1
+    end
+
+    filepath = filepath:sub(#cwd + 1 + offset, #filepath)
+  end
+
+  return filepath
+end
+
+function Path:normalize(filepath, cwd)
+  filepath = Path:make_relative(filepath, cwd)
+
+  -- Substitute home directory w/ "~"
+  filepath = filepath:gsub("^" .. path.home, '~', 1)
+
+  -- Remove double path seps, it's annoying
+  filepath = filepath:gsub(path.sep .. path.sep, path.sep)
+
+  return filepath
+end
+
+function Path:shorten(filepath)
+  if jit then
+    local ffi = require('ffi')
```

Comment:
We should only do this check once. Tj wrote the prev one as "generator" we could have a local function that works as generator as well and this function will just call the generated one.

Does that make sense?

Reply:
> Does that make sense?

Sorry... could you explain a little more?

Reply:
```lua
local this_is_a_generated_func = (function()
  if jit then
    -- the fancy ffi stuff
    return function(filepath) -- on first run this will be returned and will be the "content" of this_is_a_generated_func
      return filepath:sub(-3) -- whatever
    end
  else
    return function(filepath) -- or this will be returned. Depending on jit being available
      return filepath
    end
  end
end)() -- this will run the function immediately. When the script get read. Or so

function Path:shorten()
  return this_is_a_generated_func(parameter)
end
```

That means we only do the `jit` check and the `ffi.cdef` part only once

Reply:
That makes sense. Thanks for the help :)